_posts/2015-12-09-docker-for-development.markdown

```diff
@@ -0,0 +1,214 @@
+  command: bash -c "rm -f /data/db/mongod.lock; mongod --smallfiles --quiet --logpath=/dev/null"
+  ports: 
+  - "27017:27017"
+
+solr4:
+  image: artsy/solr4
+
+memcached:
+  image: memcached
+
+elasticsearch:
+  image: artsy/elasticsearch
+  ports:
+  - "9200:9200"
+  - "9300:9300"
+
+web:
+  extends:
+    file: common.yml
+    service: gravity
+  command: script/rails s -b 0.0.0.0 -p 80
+  ports:
+  - "80:80"
+  volumes:
+  - .:/app
+  links:
+  - elasticsearch
+  - mongodb
+  - memcached
+  - solr4
+  - mailcatcher
+
+dj:
+  extends:
+    file: common.yml
+    service: gravity
+  command: bundle exec rake jobs:work
+  volumes:
+  - .:/app
+  links:
+  - elasticsearch
+  - mongodb
+  - memcached
+  - solr4
+
+mailcatcher:
+  image: zolweb/docker-mailcatcher
+  ports:
+  - "1080:1080"
+```
+
+The command for the MongoDB section removes a lock file that can remain in place sometimes when the container is killed. Do not use that in production! We mount the local directory into the container with a `volumes:` command, so that local changes are reloaded in the running containers.
+
+Recently, [Ashkan Nasseri](https://github.com/ashkan18) began to move our delayed jobs from [delayed_job_mongoid](https://github.com/collectiveidea/delayed_job_mongoid) to [sidekiq](http://sidekiq.org/), which brings in Redis and another process that needs to run during development. Since we are using Docker, all we have to do is add a couple of new sections to our `docker-compose.yml` file:
+
+```yaml
+redis:
+  image: redis
+  ports:
+  - "6379:6379"
+
+sidekiq:
+  extends: 
+    file: common.yml
+    service: gravity
+  command: bundle exec sidekiq
+  volumes:
+  - .:/app
+  links:
+  - elasticsearch
+  - mongodb
+  - memcached
+  - solr4
+  - redis
+```
+
+and add this line to `common.yml`:
+
+```yaml
+REDIS_URL: redis://redis
+```
+
+For development which involves multiple applications in separate git repositories, we use [Dusty](http://dusty.gc.com/), which was created by [GameChanger](https://gc.com/). Some of the advantages of using Dusty include the use of `rsync` so that performance is better, and a built-in nginx proxy along with modifications to your `/etc/hosts` file so that you can more easily connect to your applications. The shared volumes approach with the `volumes:` directive can be rather slow with the default Docker Toolbox setup using VirtualBox.
+
+With Dusty, you set up services, apps, and bundles of apps with YAML files. Our MongoDB service is defined as:
+
+```yaml
+# services/mongo2.yml
+image: mongo:2.4
+volumes:
+- /persist/persistentMongo:/data/db
+entrypoint: ["sh", "-c", "rm -f /data/db/mongod.lock; mongod --smallfiles --quiet --logpath=/dev/null"]
+ports:
+- "27017:27017"
+```
+
+It's not necessary to expose the ports, but in case we want to connect directly to the MongoDB instance with the `mongo` command without shelling into a container, we need it to be available on our Docker VM's IP address.
+
+Our Gravity app's YAML file is:
```

Comment:
It may be worth specifying that `apps/gravity.yml` is used by Dusty again explicitly in this line. If you are skimming through, there are many potentially new concepts/technologies in this post and it wouldn't hurt to be overly explicative if that's a word.